[MRG] Support multi-label probability calibration

connorbrinton opened the pull request:
`CalibratedClassifierCV` now handles the calibration process in such a way that probability estimates can be calibrated for multi-label targets. Also loosens input validation requirements to better interoperate with `Pipeline`.

#### Reference Issues/PRs
Fixes #8710.

#### What does this implement/fix? Explain your changes.
Changes include (roughly in source code order):
* Looser input validation on arguments passed to wrapped classifiers (fixes #8710)
* Target classes and type are determined before cross-validation, rather than on each fold individually
* Label predictions from `CalibratedClassifierCV.predict` are obtained using `LabelBinarizer.inverse_transform`, which supports multi-label predictions
* Specialized logic in `_CalibratedClassifier` for handling binary classification problems is tidied and more thoroughly commented
* Shape of uncalibrated estimates from wrapped classifier is checked against the expected shape in `_CalibratedClassifier`
* Simplification of logic in `_CalibratedClassifier.predict_proba` along with more comments explaining what's happening
* Tests for acceptance of 1D feature arrays as input and production of valid multi-label probability predictions

#### Any other comments?
Thanks for working on scikit-learn!

connorbrinton commented:
Hi @qinhanmin2014, would you be able to review this pull request? I see that you've recently approved PRs affecting this module. If it would be more appropriate for someone else to review it, please let me know. Thanks! 🙂

cmarmo commented:
Hi @connorbrinton, I know it has been a while and I'm really sorry for that. Are you still interested in finalizing your work? If yes, do you mind synchronize with upstream? Thanks a lot for your patience!

cmarmo commented on sklearn/calibration.py:
```diff
@@ -137,6 +137,14 @@ class CalibratedClassifierCV(BaseEstimator, ClassifierMixin,
     >>> calibrated_clf.predict_proba([[-0.5, 0.5]])
     array([[0.936..., 0.063...]])
 
+    label_binarizer_ : LabelBinarizer object
+        Object used to transform labels to a binary presence matrix and
+        vice-versa.
+
+    y_type_ : str
+        Represents the type of the target data as evaluated by
+        :attr:`label_binarizer_`.
+
```
Thanks @connorbrinton! `label_binarizer_` and `y_type_` are attributes. Their description should be moved before the example. This will fix the sphinx warnings too.

connorbrinton commented:
Hi @cmarmo, thanks for following up on this PR 🙂 I rebased with upstream and fixed all of the new failures that resulted, so this PR should be ready for review again :+1:

@glemaitre Let me know if you have any questions or requests for this PR once you get the chance to review it :smile:

There is tiny amount of overlap between this PR and #17546, but the changes seem to be mostly complementary. Differences include:
* Input validation and CV fold label checks are only performed when `cv != "prefit"` in #17546
* Training input validation in this PR is loosened to allow non-numeric X (such as text input) and to not reshape 1D input to 2D (since the base estimator can do that, if needed)
* This PR modifies the call to `check_array` in `predict_proba` to match validation performed during training. In #17546 it might make sense for this validation to be conditioned on `cv != "prefit"`, to be consistent with training.

I'd be happy to rebase on it once it's merged :smile:

cmarmo commented:
#17546 has been happily merged! 🚀
@connorbrinton do you mind resolving conflicts? Thanks for your patience.

cmarmo commented:
@connorbrinton let us know if you need any help with the sync with upstream.

connorbrinton committed:
Loosen `CalibratedClassifierCV` input validation

These changes loosen `CalibratedClassifierCV`'s input validation to
accept one-dimensional and non-numeric data (such as text).

connorbrinton commented:
I'm not planning on updating this PR any more. Some of the features implemented by this PR have been implemented elsewhere, and other features haven't seemed to garner interest from maintainers. It's been a long time since there's been any movement on this PR, so there's no reason to keep it open
